### ISSUE TITLE
Reinforcement Learning can be really, really slow in tensorflow.

### ISSUE THREAD

**Issue by scroyston** [tensorflow/tensorflow#2444]
I was looking at applying DeepMind's [DDPG algorithm](http://arxiv.org/pdf/1509.02971.pdf) to the new OpenAI gym problem set, focusing on continuous control problems like: [the pendulum](https://gym.openai.com/evaluations/eval_I3x3s4DuS7CmIbpaTCs51w)

I noticed that a TensorFlow version of the algorithm was running about 10-20x slower than a theano version of the same algorithm.

[Tensorflow implementation](https://github.com/nivwusquorum/tensorflow-deepq/blob/continuous/tf_rl/controller/continuous_deepq.py)  (but with a couple bugs fixed)
[Theano implementation](https://github.com/scroyston/rllab/blob/master/rllab/algos/ddpg.py)

When I started profiling both implementations in detail, I noticed there were huge performance difference even when just comparing the forward pass of the actor neural network.

I built a quick script to just compare a forward pass of a layered net, varying the layer depth and layer breadth (code and results below), and using a "mini batch" size of 64.  (For reference, in Deepmind's DDPG paper, they used a network of 2 hidden layers, sized 400 and 300, mini batch size of 64).
Using the handy tf.RunOptions.FULL_TRACE, it showed 99% of the cost was in running the ops themselves (vs op scheduling, startup time, etc).

I then stumbled on a couple of settings that seemed to help:
    config_proto.intra_op_parallelism_threads = 1
    config_proto.inter_op_parallelism_threads = 1

That seemed to improve things a bunch, to where tensorflow was as fast in a couple instances, or "only" 2-3x slower.

Before finding the threading flags, I started looking at speeding up the individual ops. In MatMulOp I replaced the eigen matrix multiply call with a blas call.  That also seemed to help quite a bit.

Below are some chart of the performance delta vs Theano.
"EigenMultiThread" is just vanilla tensorflow. 
"EigenSingleThread" is tensorflow with intra/inter op_parallelism = 1. 
"BlasSingleThread is the same as "EigenSingleThread", with the eigen matrix multiply replaced by blas.
The data that makes up these charts can be [found here](https://docs.google.com/a/rgmadvisors.com/spreadsheets/d/1EU9xPFUqELDpRnFUOJUKCDNjm977sUC7AAM29iCwfgc/pubhtml).

![depth2](https://cloud.githubusercontent.com/assets/620227/15438353/62156fc2-1e91-11e6-9b22-5c7cf04aef91.png)
![depth4](https://cloud.githubusercontent.com/assets/620227/15438368/795d2698-1e91-11e6-979e-ea341f5d6720.png)
![depth6](https://cloud.githubusercontent.com/assets/620227/15438371/7d03b064-1e91-11e6-9589-33ccf83b926f.png)
![depth8](https://cloud.githubusercontent.com/assets/620227/15438374/80979718-1e91-11e6-8f88-ecdfeff44fef.png)

I primarily ran these tests on a mac pro (with avx).  I did compile tensorflow with the avx flag.  I also spot checked these results on a linux box and got similar results.  Blas implementation was OpenBlas

I guess my issue/question is: Is there any ongoing work to improve performance for Reinforcement Learning scenarios like these, especially on CPUs? 

I have heard DeepMind is moving completely to tensorflow, and they used/are using the new TPU chips. However, they also got a huge performance boost in runtime and error rate over GPUs by [using 16 normal cores asynchronously](http://arxiv.org/pdf/1602.01783v1.pdf) (which btw, I don't think I can set the threading options to 1 and have an [implementation](https://github.com/muupan/async-rl) still work).
### Environment info

Operating System: 
OS X 10.11.4,  3.5 GHz 6-Core Intel Xeon E5
and
Fedora 20, 12 physical cores, Xeon X5670  @ 2.93GHz

Installed version of CUDA and cuDNN: 
None

If installed from binary pip package, provide:
Tried both a release version:
pip install --upgrade https://storage.googleapis.com/tensorflow/mac/tensorflow-0.8.0-py2-none-any.whl

and from source, commit hash:
371d341c7ef22d38b80235bbcb5a6e230546781c
compile command:
bazel build -c opt --copt=-mavx //tensorflow/tools/pip_package:build_pip_package
(for the Xeon E5 machine)
### Steps to reproduce
1. I used the script below to generate the numbers
   [tf_perf_test.py.zip](https://github.com/tensorflow/tensorflow/files/275280/tf_perf_test.py.zip)

**Comment by mrry** [comment]
@benoitsteiner @rmlarsen: Benoit, Rasmus, are there other configuration options for Eigen that could help improve the performance here? Would any of the post-0.8 performance improvements be likely to help here?

**Comment by benoitsteiner** [comment]
@scroyston Haveyou tried building with the -mfma option as well ? The Mac Pro have shipped with Haswell cpus for quite some time, which support the FMA instructions. This can make a significant difference in performance.
I'll try to take a look at you testcase next week. There are a few things that we can do to improve the scalability of the code for relatively small matrix multiplication, and some of the options we're considering might help in your case.

**Comment by scroyston** [comment]
Thanks for the suggestion, unfortunately this particular mac pro doesn't seem to have fma:
[1]    13850 illegal hardware instruction

machdep.cpu.brand_string: Intel(R) Xeon(R) CPU E5-1650 v2 @ 3.50GHz
machdep.cpu.features: FPU VME DE PSE TSC MSR PAE MCE CX8 APIC SEP MTRR PGE MCA CMOV PAT PSE36 CLFSH DS ACPI MMX FXSR SSE SSE2 SS HTT TM PBE SSE3 PCLMULQDQ DTES64 MON DSCPL VMX SMX EST TM2 SSSE3 CX16 TPR PDCM SSE4.1 SSE4.2 x2APIC POPCNT AES PCID XSAVE OSXSAVE TSCTMR AVX1.0 RDRAND F16C

**Comment by Mazecreator** [comment]
I have a similar issue with Reinforcement Learning.  I am using the discrete_deepe.py from nivwusquorum [https://github.com/nivwusquorum/tensorflow-deepq/tree/master/tf_rl/controller](url).

I have a dataset with about 50,000 samples and each observation is about 80 points.  This takes about 10 Mins to processes which seems pretty slow.  I have a GPU and haven't check to see if the individual observations are being computed in that or the CPU.

**Comment by girving** [comment]
The threading issue is suspicious.  @rmlarsen, @benoitsteiner: Any thoughts as to why turning off threading would make Eigen 10x faster?

**Comment by rmlarsen** [comment]
The dependence on threading does seem suspicious. Could you try to test what happens if you only set one of the threading params to 1? In other words, run with

config_proto.intra_op_parallelism_threads = $num_cores
config_proto.inter_op_parallelism_threads = 1

or

config_proto.intra_op_parallelism_threads = 1
config_proto.inter_op_parallelism_threads = $num_cores

**Comment by Mazecreator** [comment]
Using either of those gives about the same performance as without the settings (so current and these two settings are about the same).

What is strange is I reinstalled TensorFlow R0.9 (had an earlier R0.9) without GPU support and it was about 1/2 the speed as with GPU support compiled in.  I wonder if I should open a bug report on that?  Doesn't seem to make sense as I am working with 200 x 95 batches (no 1 sample at a time) so IO shouldn't be an issue.  The newer version R0.9 is faster as I suspect it was the np.asarray() fix in the dictionary feeding.  My GPU is a 950 with compute capability of 5.2 (that is what I have compiled into TensorFlow).

I am running on a dual Xeon 6 Core (X5650) 2.66 GHz system with 72GB of RAM, but still think the GPU should improve (or not change) performance.

**Comment by Mazecreator** [comment]
Just experienced another odd event.  Here are the steps I took

1) Fresh install of "R0.9" from github
2) Configured without GPU (GPU card is install in system)
3) Built TF
4) Ran test experiment with data set ~117 seconds/iteration  (203000 observations of 95 data points in 200/batch) 
NOTE: Experiment is compiled with GPU option and loads library when TF supports GPU
5) Reconfigured with GPU Compute 5.2 (Bazel clean)
6) Ran test experiment same as in #4 above , time doubled with GPU support enabled
7) Reconfigured with GPU Compute 3.5,5.2 -> Same results as # 6
8) Reconfigured with CPU only again (Bazel clean)
9) Ran experiment, I didn't get my ~117 second/iteration, I got over 350 seconds/iteration, this was a major surprise
10) Delete TF directory, re CLONED TF from Github, compiled with CPU only, ran experiment back to 117 second without GPU support

I am not sure exactly what this experiment is indicating, but seems like when the GPU option is added, something is included in the final compiled objects that slows the matrix calculations.  I assume when I went back to the CPU only without a fresh install, some of the old GPU code was not erased but found when the experiment ran.  It didn't load the GPU libraries so it was using a different bit of code than when I first install TF without GPU support.  Hope this helps, let me know if you need more information.

**Comment by Mazecreator** [comment]
I ran a profile based upon #1824 and got the following.  I am not sure how to read this, but the GPU iteration took over 10ms where the CPU alone is <10ms.  I am running the HEAD of TensorFlow on Ubuntu 15.10 with CUDA 7.5 and cuDNN 4.  The CPU is a dual XEON 6 core processors (24 threads total) with 72 GB or RAM (DDR3).

I have a GTX 950, not the fastest, but is this related to the structure of the graph or simply the data set isn't big enough to get a benefit from the GPU given the IO overhead?

Then networks runs in this case a Batch of 200 x 189 into 5 layers with Dropout() between each layer.  The layers are 140, 120, 100, 80, and 3 as the output.

![screenshot 2016-07-10 15 08 49](https://cloud.githubusercontent.com/assets/18412448/16715534/dd719032-46b0-11e6-81cf-f0b22d96e336.png)


![screenshot 2016-07-10 15 08 58](https://cloud.githubusercontent.com/assets/18412448/16715536/e375b508-46b0-11e6-8380-bb129c380442.png)

**Comment by danijar** [comment]
Hi, what's the status of this?

**Comment by hholst80** [comment]
@scroyston very nice analysis! Can I ask how you enabled BLAS optimization for matmul?

**Comment by hholst80** [comment]
I did some [investigations](https://github.com/hholst80/tensorflow-sgemm) myself on this. I have not been able to reproduce the performance speedup in `perftest.py` in a real environment. Actually, in our "production code" we got a slowdown and I hope I can return to investigate why later on. If anyone have some battle experience to share I am happy to hear about it!

My machine is 75% loaded right now so the numbers should be taken with a grain of salt:

```
$ pip list | grep tensorflow
tensorflow (0.11.0rc0)
```

```
$ env OMP_NUM_THREADS=1 python perftest.py 
matmul 1000
0.11668629897758365
sgemm 1000
0.057338198996149004 speedup 2.0
matmul 2000
0.9684892640216276
sgemm 2000
0.352693248889409 speedup 2.7
matmul 3000
3.1248108180006966
sgemm 3000
1.1585902459919453 speedup 2.7
matmul 4000
7.574722188990563
sgemm 4000
2.3656532769091427 speedup 3.2
matmul 5000
14.732756665092893
sgemm 5000
5.034700996009633 speedup 2.9
```

`/proc/cpuinfo`
```
processor       : 0
vendor_id       : GenuineIntel
cpu family      : 6
model           : 63
model name      : Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz
stepping        : 2
microcode       : 0x36
cpu MHz         : 2802.246
cache size      : 30720 KB
physical id     : 0
siblings        : 24
core id         : 0
cpu cores       : 12
apicid          : 0
initial apicid  : 0
fpu             : yes
fpu_exception   : yes
cpuid level     : 15
wp              : yes
flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm epb tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm xsaveopt cqm_llc cqm_occup_llc dtherm ida arat pln pts
bugs            :
bogomips        : 4988.57
clflush size    : 64
cache_alignment : 64
address sizes   : 46 bits physical, 48 bits virtual
power management:
```

**Comment by girving** [comment]
@rmlarsen @benoitsteiner What's the status of this issue?

**Comment by girving** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!
